### PR TITLE
update github status to allow for drier object oriented usage

### PR DESCRIPTION
**Pull request by jebentier**
Doing this to help with dryness in pipeline files that are looking to set many statuses for the same sha

### CODE REVIEW

**Review by jadametz — COMMENTED** [review]
Can you provide a brief example of what calling this in a pipeline would look like?

**Comment by jebentier** [comment]
Yes, example new usage is
```
githubStatusChecks = new GitHubStatus(this, 'Invoca/web', env.GIT_SHA,  env.BUILD_URL, env.TOKEN)
githubStatusChecks.setStatus('jenkins-ci/production-assets-pushed', 'Assets pushed to s3://invoca-assets', 'success')
```
or you can still use the old
```
gitHubStatus([
    repoSlug:    'Invoca/web',
    sha:         env.GIT_COMMIT,
    description: 'Assets pushed to s3://invoca-assets',
    context:     'jenkins-ci/production-assets-pushed',
    targetURL:   env.BUILD_URL,
    token:       env.GITHUB_TOKEN,
    status:      'success'
])
```

**Comment by jebentier** [comment]
Yes, example new usage is
```
githubStatusChecks = new GitHubStatus(this, '<repo>', env.GIT_SHA,  env.BUILD_URL, env.TOKEN)
githubStatusChecks.setStatus('jenkins-ci/production-assets-pushed', 'Assets pushed to S3', 'success')
```
or you can still use the old
```
gitHubStatus([
    repoSlug:    '<repo>',
    sha:         env.GIT_COMMIT,
    description: 'Assets pushed to S3',
    context:     'jenkins-ci/production-assets-pushed',
    targetURL:   env.BUILD_URL,
    token:       env.GITHUB_TOKEN,
    status:      'success'
])
```

**Review by jadametz — APPROVED** [review]

**Comment by jebentier** [comment]
is there a contrib anywhere on how to apply contributions and make them public?

File: src/com/invoca/github/GitHubStatus.groovy [hunk 1]
```diff
@@ -18,11 +18,15 @@ class GitHubStatus implements Serializable {
...
+  public GitHubStatus(Script script, String repoSlug, String sha, String targetURL, String token) {
+    this.script = script
+    this.repoSlug = repoSlug
+    this.sha = sha
+    this.targetURL = targetURL
+    this.token = token
   }
 
-  void update() {
+  public void setStatus(String context, String status, String description) {
```
Comment (cmparkinson) [inline src/com/invoca/github/GitHubStatus.groovy]: `GitHubStatus` objects are meant to encapsulate an individual status.  `context` is part of the composite key that identifies each status, so it should be included in the constructor.

Nit: `public` is the default in Groovy.

**Review by cmparkinson — CHANGES_REQUESTED** [review]

**Comment by jebentier** [comment]
@cmparkinson after an offline talk i went ahead and extracted the github api portion out into it's own object, which allows for things to be dried up.  I also switched the instance method `update` to take the new status as apposed to having the status be a part of the `GitHubStatus` object.  This feel like a better contract given our discussion about the idea of getting the current status from the API.

original execution of gitHubStatus variable remains working, and drier usage of GitHubStatus looks like:
```
github_status = new GitHubStatus([
    repoSlug:    '<repo>',
    sha:         env.GIT_COMMIT,
    description: 'Assets pushed to S3',
    context:     'jenkins-ci/production-assets-pushed',
    targetURL:   env.BUILD_URL,
    token:       env.GITHUB_TOKEN
])

github_status.update('success')
github_status.update('failure')
```

**Review by cmparkinson — APPROVED** [review]

**Comment by jebentier** [comment]
@cmparkinson @jadametz am i good to merge this? and what's the release process?

**Comment by jadametz** [comment]
Looks good to merge. No formally outlined contribution guidelines at the moment. To make it available it just needs to have a tag/release. Feel free to add release notes 😄

**Pull request merged**